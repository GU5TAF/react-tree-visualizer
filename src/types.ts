import React, { ReactElement } from 'react';

export type Theme = {
  fontFamily: string;

  transitionTimingFunction: string;

  disabledColor: string;

  matchBackground: {
    wonColor: string;
    lostColor: string;
  };

  border: {
    color: string;
    highlightedColor: string;
  };

  textColor: {
    highlighted: string;
    main: string;
    dark: string;
    disabled: string;
  };

  score: {
    text: {
      highlightedWonColor: string;
      highlightedLostColor: string;
    };
    background: {
      wonColor: string;
      lostColor: string;
    };
  };

  canvasBackground: string;
};

export type Options = {
  width?: number;

  boxHeight?: number;

  canvasPadding?: number;

  spaceBetweenColumns?: number;

  spaceBetweenRows?: number;

  connectorColor?: string;

  connectorColorHighlight?: string;

  roundHeader?: {
    isShown?: boolean;
    height?: number;
    marginBottom?: number;
    fontSize?: number;
    fontColor?: string;
    backgroundColor?: string;
    fontFamily?: string;
  };

  roundSeparatorWidth?: number;

  lineInfo?: {
    separation?: number;
    homeVisitorSpread?: number;
  };

  horizontalOffset?: number;

  wonBywalkOverText?: string;

  lostByNoShowText?: string;
};

export type ComputedOptions = Options & {
  rowHeight?: number;

  columnWidth?: number;
};

export type CommonTreeProps = {
  svgWrapper?: (props: {
    bracketWidth: number;
    bracketHeight: number;
    startAt: number[];
    children: ReactElement;
  }) => React.ReactElement;

  theme?: Theme;

  options?: { style: Options };
};

export type RenderItem<T> = {
  y: number;
  id: number;
  path: number[];
  parentId?: number;
  data: T;
};

export type RenderMatrix<T> = RenderItem<T>[][];

export type Tree<T> = { data: T; children: Tree<T>[] };

export type ItemComponentProps<T> = {
  item: T;
  onMouseEnter: () => void;
  onMouseLeave: () => void;
  isCollapsed: boolean;
  toggleCollapse: () => void;
  isHighlighted: boolean;
  connectorColor: string;
  computedStyles: ComputedOptions;
};

export type GeneralTreeProps<T> = CommonTreeProps & {
  itemComponent: (props: ItemComponentProps<T>) => JSX.Element;
  currentRound?: string;
  tree: Tree<T>;
};
