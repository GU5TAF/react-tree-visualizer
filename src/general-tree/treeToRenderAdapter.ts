import { numArrayComparison } from '../utils/numbers';
import { RenderMatrix, Tree } from '../types';

/**
 * Return data should look like this
 *  [
 *  [ 1st Column ]
 *  [ 2nd Column ]
 *  [ 3rd Column ]
 *  [ 4th Column ]
 * ]
 *
 */
export function generateRenderTree<T>(
  generalTree: Tree<T>,
  collapsedPaths: number[][] = [],
  depth: number = 0,
  verticalStartingPoint = 0,
  parentId: number | undefined = undefined,
  id: number = 0,
  path: number[] = [],
  result: RenderMatrix<T> = [[]]
): {
  verticalStartingPoint: number;
  id: number;
  newResult: RenderMatrix<T>;
} {
  let currentVerticalStartingPoint = verticalStartingPoint;
  let currentResult = result;
  let currentIterationId = id;

  const { children, data, ...rest } = generalTree;
  const item = {
    ...rest,
    id: currentIterationId,
    parentId,
    y: currentVerticalStartingPoint,
    path,
    data,
  };

  currentResult[depth] = currentResult[depth]?.length
    ? [...currentResult[depth], item]
    : [item];

  if (
    children.length === 0 ||
    collapsedPaths.some(collapsedPath =>
      numArrayComparison(collapsedPath, path)
    )
  ) {
    return {
      verticalStartingPoint: currentVerticalStartingPoint + 1,
      newResult: currentResult,
      id: currentIterationId,
    };
  }

  for (let i = 0; i < children.length; i++) {
    const {
      newResult,
      verticalStartingPoint: resultingVerticalStartingPoint,
      id: resultId,
    } = generateRenderTree(
      children[i],
      collapsedPaths,
      depth + 1,
      currentVerticalStartingPoint,
      id,
      currentIterationId + 1,
      [...path, i],
      currentResult
    );
    currentIterationId = resultId;
    currentVerticalStartingPoint = resultingVerticalStartingPoint;
    currentResult = newResult;
  }

  return {
    verticalStartingPoint: currentVerticalStartingPoint,
    newResult: currentResult,
    id: currentIterationId,
  };
}
