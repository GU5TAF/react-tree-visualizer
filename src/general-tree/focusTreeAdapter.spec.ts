import { RenderMatrix, Tree } from '../types';

import { generateRenderTree } from './treeToRenderAdapter';

type TestCase = {
  generalTree?: Tree<any>;
  renderData?: RenderMatrix<any>;
  collapsedRenderData?: RenderMatrix<any>;
};
const testDictionary: TestCase = {
  generalTree: {
    data: { id: 0, ref: { current: null }, type: 0 },
    children: [
      {
        data: {
          type: 1,
          id: 4,
          ref: { current: null },
          disabled: false,
          hidden: false,
        },
        children: [
          {
            data: {
              type: 1,
              id: 14,
              ref: { current: {} },
              disabled: false,
              hidden: false,
            },
            children: [
              { data: { type: 0, id: 45, ref: { current: {} } }, children: [] },
            ],
          },
          {
            data: {
              type: 1,
              id: 15,
              ref: { current: {} },
              disabled: false,
              hidden: false,
            },
            children: [
              {
                data: {
                  scrollAxis: 1,
                  type: 1,
                  id: 17,
                  ref: { current: {} },
                  disabled: false,
                  hidden: false,
                },
                children: [
                  {
                    data: {
                      type: 0,
                      id: 22,
                      ref: { current: {} },
                    },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 23, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 24, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 25, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 26, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 27, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 28, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 29, ref: { current: {} } },
                    children: [],
                  },
                ],
              },
              {
                data: {
                  scrollAxis: 1,
                  type: 1,
                  id: 19,
                  ref: { current: {} },
                  disabled: false,
                  hidden: false,
                },
                children: [
                  {
                    data: { type: 0, id: 21, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: {
                      type: 0,
                      id: 20,
                      ref: { current: {} },
                      disabled: true,
                    },
                    children: [],
                  },
                  {
                    data: {
                      type: 0,
                      id: 30,
                      ref: { current: {} },
                    },
                    children: [],
                  },
                  {
                    data: {
                      type: 0,
                      id: 31,
                      ref: { current: {} },
                    },
                    children: [],
                  },
                  {
                    data: {
                      type: 0,
                      id: 32,
                      ref: { current: {} },
                      disabled: true,
                    },
                    children: [],
                  },
                  {
                    data: {
                      type: 0,
                      id: 33,
                      ref: { current: {} },
                      disabled: true,
                    },
                    children: [],
                  },
                  {
                    data: {
                      type: 0,
                      id: 34,
                      ref: { current: {} },
                      disabled: true,
                    },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 35, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 36, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: {
                      type: 0,
                      id: 37,
                      ref: { current: {} },
                    },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 38, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 39, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 40, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: {
                      type: 0,
                      id: 41,
                      ref: { current: {} },
                    },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 42, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: { type: 0, id: 43, ref: { current: {} } },
                    children: [],
                  },
                  {
                    data: {
                      type: 0,
                      id: 44,
                      ref: { current: {} },
                    },
                    children: [],
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
};

const biggerExample: TestCase = {
  generalTree: {
    data: { type: 0, id: 0, ref: { current: null } },
    children: [
      {
        children: [
          {
            children: [],
            data: {
              disabled: false,
              hidden: false,
              id: 2,
              ref: { current: null },
              type: 1,
            },
          },
          {
            children: [],
            data: {
              id: 3,
              disabled: false,
              hidden: false,
              ref: { current: null },
              type: 2,
            },
          },
        ],

        data: {
          id: 1,
          disabled: false,
          hidden: false,
          ref: { current: null },
          type: 1,
        },
      },
      {
        children: [
          {
            children: [],
            data: {
              disabled: false,
              hidden: false,
              id: 5,
              ref: { current: null },
              type: 1,
            },
          },
        ],
        data: {
          disabled: false,
          hidden: false,
          ref: { current: null },
          type: 2,
          id: 4,
        },
      },
    ],
  },
};

describe('given a real example general tree', () => {
  it('converts it to render data', () => {
    const { newResult } = generateRenderTree<any>(testDictionary.generalTree);

    expect(newResult).toMatchSnapshot();
  });
});

describe('given a bigger example', () => {
  it('converts it to render data', () => {
    const { newResult } = generateRenderTree<any>(biggerExample.generalTree);
    expect(newResult).toMatchSnapshot();
  });
});

describe('given a bigger example', () => {
  it('converts it to render data with collapsed children', () => {
    const { newResult } = generateRenderTree<any>(biggerExample.generalTree, [
      [0],
    ]);
    expect(newResult).toMatchSnapshot();
  });
});
