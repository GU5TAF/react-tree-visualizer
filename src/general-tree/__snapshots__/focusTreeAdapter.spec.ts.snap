// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given a bigger example converts it to render data 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "id": 0,
        "type": 0,
      },
      "id": 0,
      "parentId": undefined,
      "path": Array [],
      "uniqueId": "0",
      "y": 0,
    },
  ],
  Array [
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 1,
        "type": 1,
      },
      "id": 1,
      "parentId": 0,
      "path": Array [
        0,
      ],
      "uniqueId": "1",
      "y": 0,
    },
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 4,
        "type": 2,
      },
      "id": 4,
      "parentId": 0,
      "path": Array [
        1,
      ],
      "uniqueId": "4",
      "y": 2,
    },
  ],
  Array [
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 2,
        "type": 1,
      },
      "id": 2,
      "parentId": 1,
      "path": Array [
        0,
        0,
      ],
      "uniqueId": "2",
      "y": 0,
    },
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 3,
        "type": 2,
      },
      "id": 3,
      "parentId": 1,
      "path": Array [
        0,
        1,
      ],
      "uniqueId": "3",
      "y": 1,
    },
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 5,
        "type": 1,
      },
      "id": 5,
      "parentId": 4,
      "path": Array [
        1,
        0,
      ],
      "uniqueId": "5",
      "y": 2,
    },
  ],
]
`;

exports[`given a bigger example converts it to render data with collapsed children 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "id": 0,
        "type": 0,
      },
      "id": 0,
      "parentId": undefined,
      "path": Array [],
      "uniqueId": "0",
      "y": 0,
    },
  ],
  Array [
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 1,
        "type": 1,
      },
      "id": 1,
      "parentId": 0,
      "path": Array [
        0,
      ],
      "uniqueId": "1",
      "y": 0,
    },
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 4,
        "type": 2,
      },
      "id": 4,
      "parentId": 0,
      "path": Array [
        1,
      ],
      "uniqueId": "4",
      "y": 2,
    },
  ],
  Array [
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 2,
        "type": 1,
      },
      "id": 2,
      "parentId": 1,
      "path": Array [
        0,
        0,
      ],
      "uniqueId": "2",
      "y": 0,
    },
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 3,
        "type": 2,
      },
      "id": 3,
      "parentId": 1,
      "path": Array [
        0,
        1,
      ],
      "uniqueId": "3",
      "y": 1,
    },
  ],
]
`;

exports[`given a real example general tree converts it to render data 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "id": 0,
        "type": 0,
      },
      "id": 0,
      "parentId": undefined,
      "path": Array [],
      "uniqueId": "0",
      "y": 0,
    },
  ],
  Array [
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 4,
        "type": 1,
      },
      "id": 1,
      "parentId": 0,
      "path": Array [
        0,
      ],
      "uniqueId": "4",
      "y": 0,
    },
  ],
  Array [
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 14,
        "type": 1,
      },
      "id": 2,
      "parentId": 1,
      "path": Array [
        0,
        0,
      ],
      "uniqueId": "14",
      "y": 0,
    },
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 15,
        "type": 1,
      },
      "id": 4,
      "parentId": 1,
      "path": Array [
        0,
        1,
      ],
      "uniqueId": "15",
      "y": 1,
    },
  ],
  Array [
    Object {
      "data": Object {
        "id": 45,
        "type": 0,
      },
      "id": 3,
      "parentId": 2,
      "path": Array [
        0,
        0,
        0,
      ],
      "uniqueId": "45",
      "y": 0,
    },
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 17,
        "scrollAxis": 1,
        "type": 1,
      },
      "id": 5,
      "parentId": 4,
      "path": Array [
        0,
        1,
        0,
      ],
      "uniqueId": "17",
      "y": 1,
    },
    Object {
      "data": Object {
        "disabled": false,
        "hidden": false,
        "id": 19,
        "scrollAxis": 1,
        "type": 1,
      },
      "id": 14,
      "parentId": 4,
      "path": Array [
        0,
        1,
        1,
      ],
      "uniqueId": "19",
      "y": 9,
    },
  ],
  Array [
    Object {
      "data": Object {
        "id": 22,
        "type": 0,
      },
      "id": 6,
      "parentId": 5,
      "path": Array [
        0,
        1,
        0,
        0,
      ],
      "uniqueId": "22",
      "y": 1,
    },
    Object {
      "data": Object {
        "id": 23,
        "type": 0,
      },
      "id": 7,
      "parentId": 5,
      "path": Array [
        0,
        1,
        0,
        1,
      ],
      "uniqueId": "23",
      "y": 2,
    },
    Object {
      "data": Object {
        "id": 24,
        "type": 0,
      },
      "id": 8,
      "parentId": 5,
      "path": Array [
        0,
        1,
        0,
        2,
      ],
      "uniqueId": "24",
      "y": 3,
    },
    Object {
      "data": Object {
        "id": 25,
        "type": 0,
      },
      "id": 9,
      "parentId": 5,
      "path": Array [
        0,
        1,
        0,
        3,
      ],
      "uniqueId": "25",
      "y": 4,
    },
    Object {
      "data": Object {
        "id": 26,
        "type": 0,
      },
      "id": 10,
      "parentId": 5,
      "path": Array [
        0,
        1,
        0,
        4,
      ],
      "uniqueId": "26",
      "y": 5,
    },
    Object {
      "data": Object {
        "id": 27,
        "type": 0,
      },
      "id": 11,
      "parentId": 5,
      "path": Array [
        0,
        1,
        0,
        5,
      ],
      "uniqueId": "27",
      "y": 6,
    },
    Object {
      "data": Object {
        "id": 28,
        "type": 0,
      },
      "id": 12,
      "parentId": 5,
      "path": Array [
        0,
        1,
        0,
        6,
      ],
      "uniqueId": "28",
      "y": 7,
    },
    Object {
      "data": Object {
        "id": 29,
        "type": 0,
      },
      "id": 13,
      "parentId": 5,
      "path": Array [
        0,
        1,
        0,
        7,
      ],
      "uniqueId": "29",
      "y": 8,
    },
    Object {
      "data": Object {
        "id": 21,
        "type": 0,
      },
      "id": 15,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        0,
      ],
      "uniqueId": "21",
      "y": 9,
    },
    Object {
      "data": Object {
        "disabled": true,
        "id": 20,
        "type": 0,
      },
      "id": 16,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        1,
      ],
      "uniqueId": "20",
      "y": 10,
    },
    Object {
      "data": Object {
        "id": 30,
        "type": 0,
      },
      "id": 17,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        2,
      ],
      "uniqueId": "30",
      "y": 11,
    },
    Object {
      "data": Object {
        "id": 31,
        "type": 0,
      },
      "id": 18,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        3,
      ],
      "uniqueId": "31",
      "y": 12,
    },
    Object {
      "data": Object {
        "disabled": true,
        "id": 32,
        "type": 0,
      },
      "id": 19,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        4,
      ],
      "uniqueId": "32",
      "y": 13,
    },
    Object {
      "data": Object {
        "disabled": true,
        "id": 33,
        "type": 0,
      },
      "id": 20,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        5,
      ],
      "uniqueId": "33",
      "y": 14,
    },
    Object {
      "data": Object {
        "disabled": true,
        "id": 34,
        "type": 0,
      },
      "id": 21,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        6,
      ],
      "uniqueId": "34",
      "y": 15,
    },
    Object {
      "data": Object {
        "id": 35,
        "type": 0,
      },
      "id": 22,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        7,
      ],
      "uniqueId": "35",
      "y": 16,
    },
    Object {
      "data": Object {
        "id": 36,
        "type": 0,
      },
      "id": 23,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        8,
      ],
      "uniqueId": "36",
      "y": 17,
    },
    Object {
      "data": Object {
        "id": 37,
        "type": 0,
      },
      "id": 24,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        9,
      ],
      "uniqueId": "37",
      "y": 18,
    },
    Object {
      "data": Object {
        "id": 38,
        "type": 0,
      },
      "id": 25,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        10,
      ],
      "uniqueId": "38",
      "y": 19,
    },
    Object {
      "data": Object {
        "id": 39,
        "type": 0,
      },
      "id": 26,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        11,
      ],
      "uniqueId": "39",
      "y": 20,
    },
    Object {
      "data": Object {
        "id": 40,
        "type": 0,
      },
      "id": 27,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        12,
      ],
      "uniqueId": "40",
      "y": 21,
    },
    Object {
      "data": Object {
        "id": 41,
        "type": 0,
      },
      "id": 28,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        13,
      ],
      "uniqueId": "41",
      "y": 22,
    },
    Object {
      "data": Object {
        "id": 42,
        "type": 0,
      },
      "id": 29,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        14,
      ],
      "uniqueId": "42",
      "y": 23,
    },
    Object {
      "data": Object {
        "id": 43,
        "type": 0,
      },
      "id": 30,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        15,
      ],
      "uniqueId": "43",
      "y": 24,
    },
    Object {
      "data": Object {
        "id": 44,
        "type": 0,
      },
      "id": 31,
      "parentId": 14,
      "path": Array [
        0,
        1,
        1,
        16,
      ],
      "uniqueId": "44",
      "y": 25,
    },
  ],
]
`;
